src/sourmash/index/__init__.py

```diff
@@ -749,11 +724,11 @@ def add(self, ss, *, location=None, require_overlap=True):
         # upon insertion, count & track overlap with the specific query.
         overlap = self.orig_query_mh.count_common(ss.minhash, True)
         if overlap:
-            i = len(self.siglist)
+            md5 = ss.md5sum()
 
-            self.counter[i] = overlap
-            self.siglist.append(ss)
-            self.locations.append(location)
+            self.counter[md5] = overlap
+            self.siglist[md5] = ss
+            self.locations[md5] = location
```

Comment:
how does this affect behavior for sigs with identical md5?

Reply:
good question 😄 

This PR changes the behavior so that only the first signature added is kept. It should be identical to the previous behavior except potentially with respect to the names of signatures yielded, and it doesn't break any of the tests. It's also potentially faster in a number of circumstances relevant to #1771.

Previously this behavior was undefined, and while it's still undefined there's at least a test for it in the protocol tests - `test_counter_gather_multiple_identical_matches` - which checks that at least one of the duplicates is returned. I think that's about as formal as we should make it; the alternative would be sorting on md5sum (why that!?) or name (ok, but ...also, why that?) :)

also see [the `multipeek` issue](https://github.com/sourmash-bio/sourmash/issues/1615), which is tangentially related - what do we do about "ties" in peek? Right now we just pick the first one.

Reply:
I updated the PR description to note this, and added a test and some comments in f8e2edc1. Thanks for the prod!